rfcs/2024-05-21-ui-component-model.md

```diff
@@ -0,0 +1,248 @@
+# UI component model
+
+Authors: Gordon Brander
+
+## Overview
+
+Context: converge on a default UI component model for LLM-generated UI.
+
+In a FE context, “UI component” means a bundle of UI and behavior that together describe an island of interactivity. There are various popular component models, including React, WebComponents, Vue, etc. In keeping with common FE terminology, this rfc uses the term “component” to mean “bundle of code that describes an island of UI”. For our specific context, a UI component is probably a type of module.
+
+## Goals
+
+### Product goals
+
+- Easy for LLMs to generate
+    - Leverages patterns and/or frameworks that are widely present in the training data, or can be learned within a small context window.
+- Easy for humans to edit
+    - [Maybe it’s not even code?](https://x.com/threepointone/status/1792930000766677034) Or maybe it’s a hybrid of code and plain language?
+- Leverages familiar or established patterns for UI development
+- Conformable with existing web FE toolchains.
+
+### Technical goals
+
+- Components are **encapsulated** (P1)
+    - A component may control its child tree, but not its siblings or parent
+- Components have **inputs**, and **output UI** and **events** (P1)
+    - Components can be understood as pure-ish functions that receive props and return a view description (we may allow cheats for local component state ala hooks)
+    - Components are **black boxes**
+    - Components are decoupled and only communicate via input and output channels.
+- Component **inputs** and **outputs** are **statically-typed** (P1)
+    - E.g. via TypeScript
+    - Allows the runtime to enforce data policies on component
+- Components have **local state** (P1)
+    - State is encapsulated within component
+    - Components may pass state down to child components as input
```

Comment:
Can you spell out whether child components are internal to it or other nodes in the recipe. I assume it's the latter, but then I have two questions:
1) presumably passing state to them is just recipes wiring
2) where is the child component chosen? If it's in the component, then I think we're moving towards returning a recipe (this isn't difficult I think, and I think might look quite natural in the code). If it's in the recipe, then that's quite natural, but we are moving into a space where the UI assembly happens in recipes.